Compile error: "error: invalid initialization of reference of type ‘std::move_iterator<const long unsigned int*>::reference {aka long unsigned int&&}’ from expression of type ‘std::remove_reference<const long unsigned int&>::type {aka const long unsigned int}’"

I attempted to compile current master (as of 05 Jul 2021) on CentOS 7.9.2009 with CMake 3.17.5 and g++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44), but I get the following issue.  Is a minimum g++ version required?  v4.8.5 seems a bit dated relative to my OS release.

[ 66%] Building CXX object theta/test/CMakeFiles/theta_test.dir/theta_a_not_b_test.cpp.o
In file included from /usr/include/c++/4.8.2/bits/stl_algobase.h:67:0,
                 from /usr/include/c++/4.8.2/bits/char_traits.h:39,
                 from /usr/include/c++/4.8.2/string:40,
                 from /home/bryan/github/datasketches-cpp/common/test/catch.hpp:468,
                 from /home/bryan/github/datasketches-cpp/theta/test/theta_a_not_b_test.cpp:20:
/usr/include/c++/4.8.2/bits/stl_iterator.h: In instantiation of ‘std::move_iterator<_Iterator>::value_type&& std::move_iterator<_Iterator>::operator*() const [with _Iterator = const long unsigned int*; std::move_iterator<_Iterator>::reference = long unsigned int&&; std::move_iterator<_Iterator>::value_type = long unsigned int]’:
/usr/include/c++/4.8.2/bits/stl_algo.h:962:13:   required from ‘_OIter std::copy_if(_IIter, _IIter, _OIter, _Predicate) [with _IIter = std::move_iterator<const long unsigned int*>; _OIter = std::back_insert_iterator<std::vector<long unsigned int, std::allocator<long unsigned int> > >; _Predicate = datasketches::key_less_than<long unsigned int, long unsigned int, datasketches::trivial_extract_key>]’
/home/bryan/github/datasketches-cpp/theta/include/theta_set_difference_base_impl.hpp:49:47:   required from ‘CS datasketches::theta_set_difference_base<Entry, ExtractKey, CompactSketch, Allocator>::compute(FwdSketch&&, const Sketch&, bool) const [with FwdSketch = const datasketches::wrapped_compact_theta_sketch_alloc<std::allocator<long unsigned int> >; Sketch = datasketches::wrapped_compact_theta_sketch_alloc<std::allocator<long unsigned int> >; Entry = long unsigned int; ExtractKey = datasketches::trivial_extract_key; CompactSketch = datasketches::compact_theta_sketch_alloc<std::allocator<long unsigned int> >; Allocator = std::allocator<long unsigned int>]’
/home/bryan/github/datasketches-cpp/theta/include/theta_a_not_b_impl.hpp:37:63:   required from ‘datasketches::theta_a_not_b_alloc<Allocator>::CompactSketch datasketches::theta_a_not_b_alloc<Allocator>::compute(FwdSketch&&, const Sketch&, bool) const [with FwdSketch = const datasketches::wrapped_compact_theta_sketch_alloc<std::allocator<long unsigned int> >; Sketch = datasketches::wrapped_compact_theta_sketch_alloc<std::allocator<long unsigned int> >; Allocator = std::allocator<long unsigned int>; datasketches::theta_a_not_b_alloc<Allocator>::CompactSketch = datasketches::compact_theta_sketch_alloc<std::allocator<long unsigned int> >]’
/home/bryan/github/datasketches-cpp/theta/test/theta_a_not_b_test.cpp:186:3:   required from here
/usr/include/c++/4.8.2/bits/stl_iterator.h:963:37: error: invalid initialization of reference of type ‘std::move_iterator<const long unsigned int*>::reference {aka long unsigned int&&}’ from expression of type ‘std::remove_reference<const long unsigned int&>::type {aka const long unsigned int}’
       { return std::move(*_M_current); }
                                     ^
make[2]: *** [theta/test/CMakeFiles/theta_test.dir/theta_a_not_b_test.cpp.o] Error 1
make[1]: *** [theta/test/CMakeFiles/theta_test.dir/all] Error 2
make: *** [all] Error 2

--- a/common/include/conditional_forward.hpp
+++ b/common/include/conditional_forward.hpp
@@ -38,29 +38,41 @@ fwd_type<T1, T2> conditional_forward(T2&& value) {
 // Forward container as iterators
 
 template<typename Container>
-auto forward_begin(Container&& c) ->
-typename std::enable_if<std::is_lvalue_reference<Container>::value, decltype(c.begin())>::type
+auto forward_begin(Container&& c) -> typename std::enable_if<
+  std::is_lvalue_reference<Container>::value ||
+  std::is_same<typename std::remove_reference<Container>::type::const_iterator, decltype(c.begin())>::value,
+  decltype(c.begin())
+>::type
 {
   return c.begin();
 }
 
 template<typename Container>
-auto forward_begin(Container&& c) ->
-typename std::enable_if<!std::is_lvalue_reference<Container>::value, decltype(std::make_move_iterator(c.begin()))>::type
+auto forward_begin(Container&& c) -> typename std::enable_if<
+  !std::is_lvalue_reference<Container>::value &&
+  !std::is_same<typename std::remove_reference<Container>::type::const_iterator, decltype(c.begin())>::value,
+  decltype(std::make_move_iterator(c.begin()))
+>::type
 {
   return std::make_move_iterator(c.begin());
 }
 
 template<typename Container>
-auto forward_end(Container&& c) ->
-typename std::enable_if<std::is_lvalue_reference<Container>::value, decltype(c.end())>::type
+auto forward_end(Container&& c) -> typename std::enable_if<
+  std::is_lvalue_reference<Container>::value ||
+  std::is_same<typename std::remove_reference<Container>::type::const_iterator, decltype(c.begin())>::value,
+  decltype(c.end())
+>::type
 {
   return c.end();
 }
 
 template<typename Container>
-auto forward_end(Container&& c) ->
-typename std::enable_if<!std::is_lvalue_reference<Container>::value, decltype(std::make_move_iterator(c.end()))>::type
+auto forward_end(Container&& c) -> typename std::enable_if<
+  !std::is_lvalue_reference<Container>::value &&
+  !std::is_same<typename std::remove_reference<Container>::type::const_iterator, decltype(c.begin())>::value,
+  decltype(std::make_move_iterator(c.end()))
+>::type
 {
   return std::make_move_iterator(c.end());
 }

